community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala

```diff
@@ -36,6 +36,54 @@ trait Attribute[T] {
     array(id.x).value
   }
 
+  def isDefinedAt(id: Id): Boolean = {
+    array.size > id.x && array(id.x).hasValue
+  }
+
+  def getOrElse(id: Id, other: => T): T = {
+    if(isDefinedAt(id)) get(id) else other
+  }
+
+  def iterator: Iterator[(Id, T)] = new Iterator[(Id, T)]() {
+    private var currentId = 0
+    private var nextTup: (Id, T) = _
+
+    private def fetchNext(): Unit = {
+      nextTup = null
+      while (nextTup == null && currentId < array.size) {
+        val c = array(currentId)
+        if(c.hasValue) {
+          nextTup = (Id(currentId), c.value)
+        }
+        currentId = currentId + 1
+      }
+    }
+
+    override def hasNext = {
+      if(currentId >= array.size)
+        false
+      else  {
+        if(nextTup == null) {
+          fetchNext()
+        }
+        nextTup != null
+      }
+    }
+
+    override def next() = {
+      if(nextTup == null) {
```

Comment:
I think it leads to simpler code an less bugs if `.next()` just checks if `.hasNext()` first. This will advance the pointer for us, and we know whether to return something or not. At the moment, I think we would return null instead of throwing if you first check `.hasNext`, and ignore a false it returns and still do a `.next()`, if that makes sense.